### ISSUE TITLE
Splitting tokens into separate files

### ISSUE THREAD

**Issue by mike-engel** [drwpow/cobalt-ui#27]
As the number of tokens increase, it would be helpful to be able to split those up into separate files for maintainability. A rough proposal might look like this:

```
// tokens.json
{
  "colors": {
    "$include": "./colors.json"
  }
}
```

```
// colors.json
{
  "$type": "color",
  "white": "#FFFFFF",
  "purple": {
    "light": {
      "$value": "#F1EAFF"
    },
    "dark": {
      "$value": "#331673"
    }
  }
}
```

Which for parsing would compile that into
```
{
  "colors": {
    "$type": "color",
    "white": "#FFFFFF",
    "purple": {
      "light": {
        "$value": "#F1EAFF"
      },
      "dark": {
        "$value": "#331673"
      }
    }
  }
}
```

Alternatives could be
```
// a
// probably the simplest in code to directly replace a value, but is ambiguous and maybe hard to detect accurately
{
  "colors": "./colors.json"
}

// b
// doesn't feel great since you're conflating the top-level types for colors
{
  "colors": {
    "$type": "include",
    "$value": "./colors.json"
  }
}
```

Happy to hear your thoughts!

**Comment by mike-engel** [comment]
Thanks @drwpow, that makes sense. https://github.com/design-tokens/community-group/issues/123 seems to have stalled, but I'll hold out hope before I make some homebrew solution to combine files before running them through cobalt 😄

**Comment by drwpow** [comment]
One possible workaround that I’ve seen happen for OpenAPI specs is to make a separate one-off “bundler” that flattens multiple tokens.json schemas into one. And that happens before Cobalt scans it. This can be very crude and rudimentary, and can basically just be a spaghetti script that spits out a single tokens.json file that’s ignored from Git. You could also shortcut it because you know exactly what you do/don’t need.

From my work in bundlers, I do know that dealing with conflicts/overrides/outlining what are valid and invalid subschemas (e.g. can a single hex color be a subschema?) you delve into the realm of differences of opinion quite quickly. And there are so many things people _want_ to do that you don’t even think should be possible let alone a good idea 😄. But anyways, again, all that to say that **solving a single example of a multi-file schema** is actually pretty easy. But solving that problem holistically for most users is complex, and is far more community alignment than a technical challenge.

**Comment by drwpow** [comment]
@mike-engel just shipped support for this in the latest version—you can just pass an array in. With the caveat again that tokens will get overridden (no warning for now, but if we want to warn or fail we can)

**Comment by mike-engel** [comment]
Awesome, thanks @drwpow. Going to remove our custom solution in favor of this, and will let you know if we run into issues 😄

**Comment by drwpow** [comment]
Oh that makes sense. I’d probably call it `co bundle`, and that’d be a separate command from `build`.

In a sense, `tokens.json` isn’t an _output_ from a build command; it’s the source of truth that powers the build command. And source files shouldn’t ever be touched by an automated tool. However, you may be taking once source of truth (or multiple), and then transforming it, and _then_ running another build command on that compiled source of truth. That way there’s integrity that the bundled file produces the exact artifacts you want, and isn’t just some untested thing.

Seems like an extra step, I know, but I’m taking all this from the Redoc CLI for REST OpenAPI schemas which I can’t say enough good things about. Been following this exact process for OpenAPI schemas (sources of truth are split up, but the CLI bundles them, and other tools import the bundle) and it’s been great.

**Comment by mike-engel** [comment]
`co bundle` makes sense to me, and would be perfect for what we need, which is just a token file compiler 😄

**Comment by drwpow** [comment]
Shipped a `co bundle` command in `1.4.0`!
